### PR TITLE
Hook stops without poststop from runc

### LINKED ISSUE #15: oci-seccomp-bpf-hook should be able to tell from bpf that the pid 1 of container exited
The hook should not have to rely on runc sending the poststop hook and sending a signal across.

Isn't there something BPF would tell us?

### PR DESCRIPTION
**Pull request by weirdwiz**
The hook can check if the container has exited and exits if the container has exited. 

Fix #15

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1 +1,2 @@
 /bin/
+/docs/*.[158]
```
Comment (rhatdan) [inline .gitignore]: Add New Line

File: Makefile [hunk 2]
```diff
@@ -87,13 +87,11 @@ install:
...
-	sed -i 's|HOOK_BIN_DIR|$(HOOK_BIN_DIR)|g' ${DESTDIR}$(HOOK_DIR)/oci-seccomp-bpf-hook-stop.json
 	$(MAKE) -C docs install
 
 clean: ## Clean artifacts
 	$(MAKE) -C docs clean
 	rm -rf \
 		bin
 	find . -name \*~ -delete
-	find . -name \#\* -delete
+	find . -name \#\* -delete
```
Comment (rhatdan) [inline Makefile]: Add new line

File: oci-seccomp-bpf-hook.go [hunk 3]
```diff
@@ -97,11 +101,24 @@ int enter_trace(struct tracepoint__raw_syscalls__sys_enter *args)
     if (*parent_inum != inum)
     {
         return 0;
-    }
-
+	}
```
Comment (rhatdan) [inline oci-seccomp-bpf-hook.go]: Bad indent?

**Comment by rhatdan** [comment]
@vrothberg PTAL

**Comment by rhatdan** [comment]
Man page needs to be updated as well.

File: oci-seccomp-bpf-hook.go [hunk 4]
```diff
@@ -33,6 +34,8 @@ type event struct {
 	ID uint32
 	// Command which makes the syscall
 	Command [16]byte
+	// Check exit
```
Comment (vrothberg) [inline oci-seccomp-bpf-hook.go]: Please add a more verbose comment. What's the semantics of this field? How and when is it used?

File: oci-seccomp-bpf-hook.go [hunk 5]
```diff
@@ -63,7 +66,8 @@ struct syscall_data
 	// the syscall number
 	u32 id;
 	// command which is making the syscall
-    char comm[16];
+	char comm[16];
+	int exit;
```
Comment (vrothberg) [inline oci-seccomp-bpf-hook.go]: Please add a comment to the field.

File: oci-seccomp-bpf-hook.go [hunk 6]
```diff
@@ -97,11 +101,24 @@ int enter_trace(struct tracepoint__raw_syscalls__sys_enter *args)
     if (*parent_inum != inum)
     {
         return 0;
-    }
-
+	}
+	
+	data.exit = 0;
```
Comment (vrothberg) [inline oci-seccomp-bpf-hook.go]: Indentation seems of here.

File: oci-seccomp-bpf-hook.go [hunk 7]
```diff
@@ -234,18 +251,26 @@ func runBPFSource(pid int, profilePath string, log *logrus.Logger) error {
...
-		return fmt.Errorf("unable to load tracepoint err:%q", err.Error())
+	if err := m.AttachTracepoint("raw_syscalls:sys_enter", enterTrace); err != nil {
+		return fmt.Errorf("unable to load enter_trace err:%q", err.Error())
+	}
+	if err := m.AttachTracepoint("sched:sched_process_exit", checkExit); err != nil {
+		return fmt.Errorf("unable to load check_exit err:%q", err.Error())
 	}
 
-	// send a signal to the parent process to indicate the compilation has been completed
+	//send a signal to the parent process to indicate the compilation has been completed
```
Comment (vrothberg) [inline oci-seccomp-bpf-hook.go]: Space between comment `//` and `send`

File: oci-seccomp-bpf-hook.go [hunk 8]
```diff
@@ -258,9 +283,10 @@ func runBPFSource(pid int, profilePath string, log *logrus.Logger) error {
 		return fmt.Errorf("unable to init perf map err:%q", err.Error())
 	}
 
-	sig := make(chan os.Signal, 1)
-	signal.Notify(sig, os.Interrupt)
 	reachedPRCTL := false // Reached PRCTL syscall
+
+	wg.Add(1)
```
Comment (vrothberg) [inline oci-seccomp-bpf-hook.go]: Please add a comment why we're doing that.

File: oci-seccomp-bpf-hook.go [hunk 9]
```diff
@@ -270,24 +296,29 @@ func runBPFSource(pid int, profilePath string, log *logrus.Logger) error {
...
 			}
 		}
+		wg.Done()
 	}()
-	log.Println("PerfMap Start")
 	perfMap.Start()
-	<-sig
-	log.Println("PerfMap Stop")
+	log.Println("PerfMap Start")
+	wg.Wait()
```
Comment (vrothberg) [inline oci-seccomp-bpf-hook.go]: Please add a comment why and what we are waiting for.

File: Makefile (hunk 2 above)
Comment (weirdwiz) [reply Makefile]: fixed :D

**Comment by weirdwiz** [comment]
The issues are fixed :D

File: oci-seccomp-bpf-hook.go [hunk 10]
```diff
@@ -59,15 +59,15 @@ BPF_HASH(parent_namespace, u64, unsigned int);
...
-	int exit;
+struct syscall_data {
+    // PID of the process
+    u32 pid;
+    // the syscall number
+    u32 id;
+    // command which is making the syscall
+    char comm[16];
+    // a flag to check if the parent process is running
+    int running;
```
Comment (vrothberg) [inline oci-seccomp-bpf-hook.go]: I don't like the naming and description of this field. It's a bit misleading as when we're setting it to 1, the "parent" is actually the current process.

I suggest calling `stopTracing` and make it a bool. This seems more straight forward and easier to document (i.e., when are we stopping to trace?).

**Comment by vrothberg** [comment]
LGTM, @weirdwiz please squash the commits into one. Then we can merge.

**Comment by weirdwiz** [comment]
Done!

**Pull request merged**